Configurable mod key

notpeelz opened the pull request:
Addresses #197 and (indirectly) #682

1. At first, I had two settings: `mod-key` and `nested-mod-key`. I tried preserving the existing code structure, but unfortunately it made the code overly complicated.
2. I dropped the `nested-mod-key` setting in 40df1356dae07c6689ad9c125f5cb3a8a518c7e5 as an attempt to simplify the logic.
3. In order to restore the "alternate mod key" behavior for the winit backend, I added an environment variable to override the mod key without altering the config file: c87287b436b3d9b8bf1baa69016f90e079f7dd3e

Question: I didn't know where to add the setting. Is `input { mod-key "super"; }` fine?

Feedback welcome :)

YaLTeR commented on niri-config/src/lib.rs:
```diff
@@ -3022,13 +3086,28 @@ where
                 }
                 Ok(bind) => {
-                    if seen_keys.insert(bind.key) {
+                    let Some(mod_key) = &mod_key else {
+                        // On first config pass, we don't know what the ModKey is, so don't check
+                        // for bind collisions.
+                        binds.push(bind);
+                        continue;
+                    };
+
+                    let mut key = bind.key;
+                    if key.modifiers.contains(Modifiers::COMPOSITOR) {
+                        key.modifiers.remove(Modifiers::COMPOSITOR);
+                        key.modifiers.insert(mod_key.0);
+                    }
+
+                    if seen_keys.insert(key) {
```
Why all this new logic? It works fine now (even with mod key being different between host and nested), so why do this with configurable mod key?

notpeelz replied:
While the current setup works, there's an edge-case bug I'd like to address. For example, with a config like this:

```
input {
  mod-key "super"
}

binds {
  Mod+Z { spawn "kitty"; }
  Super+Z { show-hotkey-overlay; }
}
```

The parser can't detect conflicts between Mod+Z and Super+Z unless it knows what `Mod` maps to.

Here's the approach I propose:

1. Parse the config once (without checking for conflicting binds).
2. After parsing, validate the binds with the the `mod-key` it just extracted from the config.
3. If duplicates exist, re-parse with `ModKey` passed via the context to generate a `DecodeError` (leveraging knuffel's error formatting).

This way, we avoid custom error handling and ensure conflicts are reported accurately by the parser.
What do you think?

notpeelz replied:
> and it shouldn't

Why not? In this scenario, `Mod+Z` and `Super+Z` clearly conflict, so I'd expect it to report an error.

> You don't want your config to suddenly break when you open nested for example.

You can use `NIRI_OVERRIDE_MOD` for developing/testing niri. Do you think that's good enough?

YaLTeR replied:
> You can use NIRI_OVERRIDE_MOD for developing/testing niri. Do you think that's good enough?

See my comment in the other thread on why I don't think that's a good solution.

notpeelz replied:
What about the conflicts question? My current implementation allows multiple modifiers in `mod-key`:

```
input {
  mod-key "ctrl+shift"
}

binds {
  Mod+Z { spawn "kitty"; }
  // this conflicts with the bind above
  Ctrl+Shift+Z { show-hotkey-overlay; }
}
```

Should I simplify the implementation and only allow a single modifier for `mod-key`?

YaLTeR replied:
I think you should allow only single if that's fine. And I think I would leave conflict resolution out, at least for this PR.

notpeelz replied:
Sounds good. I'll take another crack at it later.

YaLTeR commented on src/main.rs:
```diff
@@ -93,14 +94,28 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
     // Set a better error printer for config loading.
     niri_config::set_miette_hook().unwrap();
 
+    let mod_override = {
+        let s = std::env::var("NIRI_OVERRIDE_MOD").ok();
+        std::env::remove_var("NIRI_OVERRIDE_MOD");
+        s.and_then(|x| match ModKey::from_str(&x) {
+            Ok(x) => Some(x),
+            Err(err) => {
+                warn!("invalid value for NIRI_OVERRIDE_MOD: {err}");
+                None
+            }
+        })
+    };
+
+    let config_opts = ConfigOpts { mod_override };
+
```
I don't think this is necessary at all? Just leave a `mod-key` setting. For nested mod key, you could have a `mod-key-nested` like you say you originally had, and default it to a simple condition: if host mod key is alt then use super nested mod, otherwise use alt nested mod.

notpeelz replied:
> if host mod key is alt then use super nested mod, otherwise use alt nested mod.

I'm not sure I understand what you mean. Can you clarify?

From my perspective, exposing the "nested mod" as a setting doesn't seem necessary since the winit backend is primarily for development and testing.

Adding a `mod-key-nested` setting would require passing a `mod_key` field/parameter to several methods and structs, which I'd prefer to avoid. It also complicates the [`HotkeyOverlay` update logic](https://github.com/notpeelz/niri/commit/af1fdbe8d881cb80602af5f65d4ae1ccabca9b2c#diff-7d6e974fc702a349992658e868735716105361af2179217551bfe6d09be4ad9dR72-R75) (and any other structs that depend on the mod key), which I find a bit cumbersome.

In 40df1356dae07c6689ad9c125f5cb3a8a518c7e5, you can see how removing the backend dependency simplifies the logic. More importantly, this keeps `Config` as the single source of truth.

When hacking on niri, I usually run `NIRI_CONFIG=$PWD/config.kdl cargo run` and set `mod-key` to something that doesn't conflict with my primary compositor. For convenience, I added `NIRI_OVERRIDE_MOD` as an alternative.

YaLTeR replied:
Running `niri` should *just work* for most common host compositor setups. No environment variables, no custom config. Most commonly, the host mod key will be Super, so nested niri should default to Alt. If you override your host niri key to Alt, then it likely makes sense to default to Super for nested niri.

You can leave it non-configurable, but it will still need to change to Alt on its own when running on nested. People who first-try niri will not set any environment variables or change any config.

> When hacking on niri, I usually run NIRI_CONFIG=$PWD/config.kdl cargo run and set mod-key to something that doesn't conflict with my primary compositor.

When hacking on niri I don't set any environment variables and don't use a custom config. It should not be required.

notpeelz replied:
> Running niri should just work for most common host compositor setups. No environment variables, no custom config.

Gotcha, I'll keep that in mind for my next iteration of this PR.

> Most commonly, the host mod key will be Super, so nested niri should default to Alt. If you override your host niri key to Alt, then it likely makes sense to default to Super for nested niri.

My use case may be a bit uncommon. I have caps-lock mapped to Mod5/ISO_Level3_Shift, which I use as my `mod-key`. I guess it should map to Alt in this scenario?

> You can leave it non-configurable, but it will still need to change to Alt on its own when running on nested. People who first-try niri will not set any environment variables or change any config.

Got it.

notpeelz commented:
I rebased my changes and got rid of multi-modifier support for `ModKey`.

Questions:
- I don't remember why I renamed the type from `CompositorMod` to `ModKey`. What do you prefer?
- I changed the displayed names in hotkey_overlay for `ISO_LevelX_Shift` to `Mod5` and `Mod3`. `ISO_LevelX_Shift` is just way too long. Is that okay?

YaLTeR commented on niri-config/src/lib.rs:
```diff
@@ -3364,6 +3391,52 @@ where
     }
 }
 
+impl<S: knuffel::traits::ErrorSpan> knuffel::DecodeScalar<S> for ModKey {
```
Why do you need a custom DecodeScalar? knuffel can use FromStr on its own with `unwrap(argument, str)`

YaLTeR commented on src/input/mod.rs:
```diff
@@ -2913,6 +2891,7 @@ impl State {
         // We're using touch, hide the pointer.
         self.niri.pointer_hidden = true;
     }
+
```
unrelated change

YaLTeR commented on src/niri.rs:
```diff
@@ -591,13 +591,15 @@ impl State {
             Backend::Tty(tty)
         };
 
+        let mod_key = get_mod_key(&*config.borrow(), &backend);
```
Doesn't seem like you need to pass this in? Both config and backend are passed in, so you can get it inside.

YaLTeR commented on src/niri.rs:
```diff
@@ -1227,14 +1229,17 @@ impl State {
             self.niri.mods_with_finger_scroll_binds =
-                mods_with_finger_scroll_binds(self.backend.mod_key(), &config.binds);
+                mods_with_finger_scroll_binds(new_mod_key, &config.binds);
+            self.niri
+                .hotkey_overlay
+                .on_hotkey_config_updated(new_mod_key);
```
Probably don't move this statement to avoid unnecessary diff

YaLTeR commented on src/niri.rs:
```diff
@@ -1813,6 +1818,17 @@ impl State {
     }
 }
 
+pub fn get_mod_key(config: &Config, backend: &Backend) -> ModKey {
+    match backend {
+        Backend::Winit(_) => config
+            .input
+            .mod_key_nested
+            .or(config.input.mod_key)
+            .unwrap_or(ModKey::Alt),
+        Backend::Tty(_) | Backend::Headless(_) => config.input.mod_key.unwrap_or(ModKey::Super),
+    }
+}
```
I think this can remain a function on the Backend then?

YaLTeR reviewed (changes requested):
Looks much better now.

> I don't remember why I renamed the type from CompositorMod to ModKey. What do you prefer?

While I don't much like the extra diff, ModKey matches the new input setting and it also brought some multiple lines back to one, so let's leave it.

> I changed the displayed names in hotkey_overlay for ISO_LevelX_Shift to Mod5 and Mod3. ISO_LevelX_Shift is just way too long. Is that okay?

That's fine I guess.

notpeelz commented:
Should be good now. Let me know if there's anything else that needs changing.

YaLTeR reviewed (approved):
Did some cosmetic changes and wrote docs, good to go now.

YaLTeR commented:
Thanks!
